### PR TITLE
Add HTTP SD recommendations

### PR DESCRIPTION
**Pull request by roidelapluie**
Signed-off-by: Julien Pivotto <roidelapluie@inuits.eu>

### CODE REVIEW

File: content/docs/instrumenting/http_sd.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+request contains a `X-Prometheus-Refresh-Interval-Seconds` HTTP header with the
+refresh interval.
+
+The SD endpoint must answer with an HTTP 200 response, with the HTTP Header
+`Content-Type: application/json`. The answer must be UTF-8 formatted.
+In the case when no targets are returned, an HTTP 200 must also be emitted, with
+an empty list `[]`.
+
+It is expected that the whole list of targets is returned on every scrape. There
+is no client-side caching mechanism. A prometheus instance does not send its
```
Comment (SuperQ) [inline content/docs/instrumenting/http_sd.md:40]: Does this imply that a failure (4xx, 5xx, TCP RST) will empty any existing discovery data? If any failure results in no targets, we should probably call this out.

Comment (roidelapluie) [reply content/docs/instrumenting/http_sd.md:40]: It should keep the previous targets set. I will update.

Comment (SuperQ) [reply content/docs/instrumenting/http_sd.md:40]: Rather than "no client-side cache", I would say "It does not support incremental updates".

Comment (SuperQ) [reply content/docs/instrumenting/http_sd.md:40]: Also "It does not store data between restarts"

**Comment by roidelapluie** [comment]
@SuperQ  I have improved this based on your feedback

File: content/docs/instrumenting/http_sd.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
+---
+title: HTTP SD
+sort_rank: 7
+---
+
+# Writing HTTP Service Discovery
+
+Prometheus provides a generic [HTTP Service Discovery](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#http_sd_config),
+that enables to discover targets over an HTTP endpoint.
```
Comment (RichiH) [inline content/docs/instrumenting/http_sd.md:9]:
```suggestion
that enables it to discover targets over an HTTP endpoint.
```

File: content/docs/instrumenting/http_sd.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+The HTTP Service Discovery is complimentary to the supported service
+discoveries, and is an alternative to the [File-based Service Discovery](https://prometheus.io/docs/guides/file-sd/#use-file-based-service-discovery-to-discover-scrape-targets).
```
Comment (RichiH) [inline content/docs/instrumenting/http_sd.md:12]:
```suggestion
discovery mechanisms, and is an alternative to [File-based Service Discovery](https://prometheus.io/docs/guides/file-sd/#use-file-based-service-discovery-to-discover-scrape-targets).
```

File: content/docs/instrumenting/http_sd.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+| ---- | ------- | ------- |
+| Event Based | Yes, via inotify | No |
+| Update frequency | Instant, thanks to inotify | Following refresh_interval |
+| Format | Yaml or JSON | JSON |
+| Transport | Local file | HTTP/HTTPS |
+| Security | File-Based security | TLS, Basic auth, Authorization header, OAuth2 |
+
+## Requirements of HTTP SD endpoints
+
+The HTTP SD endpoints should have been written specifically for Prometheus.
```
Comment (RichiH) [inline content/docs/instrumenting/http_sd.md:28]: That's unclear.

File: content/docs/instrumenting/http_sd.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+The response is consumed as is, unmodified. On each refresh interval (default: 1
+minute), Prometheus will do a GET request to the HTTP SD endpoint. The GET
+request contains a `X-Prometheus-Refresh-Interval-Seconds` HTTP header with the
+refresh interval.
+
+The SD endpoint must answer with an HTTP 200 response, with the HTTP Header
+`Content-Type: application/json`. The answer must be UTF-8 formatted.
+In the case when no targets are returned, an HTTP 200 must also be emitted, with
```
Comment (RichiH) [inline content/docs/instrumenting/http_sd.md:36]:
```suggestion
If no targets should be transmitted, HTTP 200 must also be emitted, with
```

File: content/docs/instrumenting/http_sd.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+an empty list `[]`.
+
+If an error occurs while fetching the targets list, the current targets list is
+kept. The targets list is not saved across restart.
```
Comment (RichiH) [inline content/docs/instrumenting/http_sd.md:40]:
```suggestion
Prometheus caches target lists. If an error occurs while fetching an updated targets list, Prometheus keeps using the current targets list. The targets list is not saved across restart.
```

File: content/docs/instrumenting/http_sd.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+The whole list of targets must be returned on every scrape. There is no support
+for incremental update. A prometheus instance does not send its hostname and it
```
Comment (RichiH) [inline content/docs/instrumenting/http_sd.md:43]:
```suggestion
for incremental updates. A Prometheus instance does not send its hostname and it
```

File: content/docs/instrumenting/http_sd.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+is not possible for a SD endpoint to know if the SD requests is the first one
+after a restart or not.
+
+The URL to the HTTP SD is not considered secret. The authentication, and any API
+keys should be passed with the appropriate authentication mechanisms. Prometheus
+supports TLS authentication, basic authentication, oauth2, and authorization
```
Comment (RichiH) [inline content/docs/instrumenting/http_sd.md:49]:
```suggestion
supports TLS authentication, basic authentication, OAuth2, and authorization
```

**Review by RichiH — CHANGES_REQUESTED** [review]

File: content/docs/instrumenting/http_sd.md (hunk 4 above)
Comment (roidelapluie) [reply content/docs/instrumenting/http_sd.md:28]: What I try to say really is that we expect this to work only for endpoints crafted for this purpose. We don't aim to support e.g. file servers that would not provide appropriate content types etc...

Comment (roidelapluie) [reply content/docs/instrumenting/http_sd.md:28]: How can we better phrase it?

**Comment by roidelapluie** [comment]
Thanks @RichiH , ready for a next round of reviews.

File: content/docs/instrumenting/http_sd.md [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+| Transport | Local file | HTTP/HTTPS |
+| Security | File-Based security | TLS, Basic auth, Authorization header, OAuth2 |
+
+## Requirements of HTTP SD endpoints
+
+If you implement an HTTP SD endpoint, here is a few requirements you should be
+aware of.
+
+The response is consumed as is, unmodified. On each refresh interval (default: 1
+minute), Prometheus will do a GET request to the HTTP SD endpoint. The GET
```
Comment (RichiH) [inline content/docs/instrumenting/http_sd.md:32]:
```suggestion
minute), Prometheus will perform a GET request to the HTTP SD endpoint. The GET
```

File: content/docs/instrumenting/http_sd.md [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+request contains a `X-Prometheus-Refresh-Interval-Seconds` HTTP header with the
```
Comment (RichiH) [inline content/docs/instrumenting/http_sd.md:33]: Why? What do you suggest implementers do with it?

File: content/docs/instrumenting/http_sd.md [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+| Item | File SD | HTTP SD |
+| ---- | ------- | ------- |
+| Event Based | Yes, via inotify | No |
+| Update frequency | Instant, thanks to inotify | Following refresh_interval |
+| Format | Yaml or JSON | JSON |
+| Transport | Local file | HTTP/HTTPS |
+| Security | File-Based security | TLS, Basic auth, Authorization header, OAuth2 |
+
+## Requirements of HTTP SD endpoints
```
Comment (RichiH) [inline content/docs/instrumenting/http_sd.md:26]: I wonder if we should call this v1 or 1.0.0 and use RFC 2119 language

**Review by RichiH — CHANGES_REQUESTED** [review]

File: content/docs/instrumenting/http_sd.md (hunk 10 above)
Comment (roidelapluie) [reply content/docs/instrumenting/http_sd.md:33]: X-Prometheus-Refresh-Interval-Seconds is 1, they might cache on their side. I don't want to call this explicitly, I just want to guarantee that we will keep this field in the future.

File: content/docs/instrumenting/http_sd.md (hunk 11 above)
Comment (roidelapluie) [reply content/docs/instrumenting/http_sd.md:26]: We can version later. At this point, this is not even released. I would refrain from versioning or being too strict here. It is almost rfc2119, we could just turn lowercases into uppercases later on.

File: content/docs/instrumenting/http_sd.md [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+
+Prometheus caches target lists. If an error occurs while fetching an updated
+targets list, Prometheus keeps using the current targets list. The targets list
+is not saved across restart.
+
+The whole list of targets must be returned on every scrape. There is no support
+for incremental updates. A Prometheus instance does not send its hostname and it
+is not possible for a SD endpoint to know if the SD requests is the first one
+after a restart or not.
+
```
Comment (TimSimmons) [inline content/docs/instrumenting/http_sd.md:49]: random (probably irrelevant) question, need there be any consideration for sorting the results of the HTTP endpoint? would it cause unnecessary updates to the list of targets (potentially resetting scrape timers or something unintended) if the results "changed" by being in a different order?

File: content/docs/instrumenting/http_sd.md (hunk 11 above)
Comment (roidelapluie) [reply content/docs/instrumenting/http_sd.md:26]: I propose we do RFC2119 and versioning in the compliance repository.

File: content/docs/instrumenting/http_sd.md [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+The response is consumed as is, unmodified. On each refresh interval (default: 1
+minute), Prometheus will perform a GET request to the HTTP SD endpoint. The GET
+request contains a `X-Prometheus-Refresh-Interval-Seconds` HTTP header with the
+refresh interval.
+
+The SD endpoint must answer with an HTTP 200 response, with the HTTP Header
+`Content-Type: application/json`. The answer must be UTF-8 formatted.
+If no targets should be transmitted, HTTP 200 must also be emitted, with
+an empty list `[]`.
```
Comment (RichiH) [inline content/docs/instrumenting/http_sd.md:39]:
```suggestion
an empty list `[]`. The target list is unordered.
```

**Review by RichiH — CHANGES_REQUESTED** [review]
@TimSimmons is right in https://github.com/prometheus/docs/pull/1975#discussion_r651926479 -- easy fix.

File: content/docs/instrumenting/http_sd.md [hunk 14]
```diff
@@ -36,7 +36,7 @@ refresh interval.
 The SD endpoint must answer with an HTTP 200 response, with the HTTP Header
 `Content-Type: application/json`. The answer must be UTF-8 formatted.
 If no targets should be transmitted, HTTP 200 must also be emitted, with
-an empty list `[]`.
+an empty list `[]`. The list of unordered.
```
Comment (RichiH) [inline content/docs/instrumenting/http_sd.md:39]: I would call it target list, but nm.

`s/of/is/` and good to merge.

**Review by RichiH — CHANGES_REQUESTED** [review]

File: content/docs/instrumenting/http_sd.md (hunk 14 above)
Comment (roidelapluie) [reply content/docs/instrumenting/http_sd.md:39]: What about Lists are unordered ?

Because this is a list of  maps that contain a target list.

Comment (RichiH) [reply content/docs/instrumenting/http_sd.md:39]: We're on the long tail of the discussion; "Target lists are unordered" would be even more generic; I am fine with any.

**Comment by RichiH** [comment]
Thanks!

**Comment by juliusv** [comment]
Hi,

Two notes:

* Should this have been merged already, given that HTTP SD is not yet released?
* I'm unsure as to why this is placed under the "Instrumenting" section of our docs page? That section in my mind should only be about adding metrics to things / getting metrics out of things. Would it fit somehow under "Operating" or "Best Practices", or do we need to find another place to fit it in?

**Comment by roidelapluie** [comment]
> Hi,
> 
> Two notes:
> 
> * Should this have been merged already, given that HTTP SD is not yet released?

Yes, several people are already working on HTTP SD instrumentation.

> * I'm unsure as to why this is placed under the "Instrumenting" section of our docs page? That section in my mind should only be about adding metrics to things / getting metrics out of things. Would it fit somehow under "Operating" or "Best Practices", or do we need to find another place to fit it in?

My reasoning:

I have put it into instrumenting because it is to instrument other software to offer SD capabilities. I see instrumenting as a section about development.

This is not best practices, as the rules are strict, and if you do not follow them, your SD would not work.

This is not operating, even if the list of software that supports it would go under operating. This HTTP SD development section is more about development, therefore not the same public.

**Comment by juliusv** [comment]
> Yes, several people are already working on HTTP SD instrumentation.

At the moment it contains broken links though (the one to the configuration section on the config page, which doesn't contain the `http_sd_config` yet because 2.28 is not yet released.

> I have put it into instrumenting because it is to instrument other software to offer SD capabilities. I see instrumenting as a section about development.

Instrumenting is about adding metrics to things, and the section has always been about that only (see also https://en.wikipedia.org/wiki/Instrumentation). So I am almost 100% sure that it does not fit into that section.

Maybe we need to come up with a completely new section with information about building non-instrumentation integrations. Or for now, just have an "HTTP SD" page under the existing "Prometheus" point? That has already become kind of a catch-all for Prometheus-server-specific docs, and at least it would fit in nicely in that way.

**Comment by roidelapluie** [comment]
We can't have in under Prometheus as I'd rather have this independant of the Prometheus version.

Best practice seems the best fit then for now?

**Comment by juliusv** [comment]
> We can't have in under Prometheus as I'd rather have this independant of the Prometheus version.

How would it hurt to have it in the Prometheus docs though? If it's version-independent, all that means is that the content won't change between versions. But we do the same for "Federation" and some other points which are pretty version-independent, but thematically tied to the Prometheus server (like HTTP SD).

**Comment by RichiH** [comment]
Versioned makes sense, either way. Putting it under the versioning of Prometheus is the quickest way.

**Comment by roidelapluie** [comment]
Okay. @juliusv would you have the time to do this before the release?
Would be nice to also have an entry here: https://github.com/prometheus/docs/blob/master/content/_redirects

**Comment by juliusv** [comment]
@roidelapluie Yeah. Since it's just docs, do you think it's ok to just put it into the full release right after rc.0, without doing an rc.1?

**Comment by roidelapluie** [comment]
Yes ofc

**Comment by juliusv** [comment]
Thought so. Done in https://github.com/prometheus/docs/pull/1976 and https://github.com/prometheus/prometheus/pull/8972.

**Pull request merged**